keycastr/KCAppController.m

```diff
@@ -397,46 +466,74 @@ -(void) restartPanel:(NSAlert*)alert closedWithCode:(NSModalResponse)returnCode
 
 -(void) changeIconPreference:(id)sender
 {
-	int newIconPref = [[NSUserDefaults standardUserDefaults] integerForKey:kKCPrefDisplayIcon];
-	if ((newIconPref & 0x02) != (_startupIconPreference & 0x02))
-	{
-		NSString* displayMessage = (_startupIconPreference & 0x02)
-			? @"In order to hide the dock icon, KeyCastr must be restarted."
-			: @"In order to show the dock icon, KeyCastr must be restarted.";
-
-        NSAlert *alert = [[NSAlert alloc] init];
-        alert.messageText = @"Restart required";
-        alert.informativeText = displayMessage;
-        [alert addButtonWithTitle:@"Restart Now"];
-        [alert addButtonWithTitle:@"Restart Later"];
-        [alert beginSheetModalForWindow:preferencesWindow completionHandler:^(NSModalResponse returnCode) {
-            [self restartPanel:alert closedWithCode:returnCode];
-        }];
-	}
-	
-	if (newIconPref & 0x01)
-		[self createStatusItem];
-	else
-	{
-		if (_startupIconPreference & 0x02)
-			[self deleteStatusItem];
-		else
-		{
-			// On startup, we did not have an icon in the dock.
-			// Don't remove the menu bar item, because if we did,
-			// KeyCastr would be inaccessible.
-		}
-	}
-	
-	NSTask* task = [NSTask launchedTaskWithLaunchPath:@"/usr/bin/defaults" arguments:[NSArray arrayWithObjects:
-		@"write",
-		[NSString stringWithFormat:@"%@/Contents/Info", [[NSBundle mainBundle] bundlePath]],
-		@"LSUIElement",
-		((newIconPref & 0x02)
-			? @"0"
-			: @"1"),
-		nil]];
-	[task waitUntilExit];
+    // sent from the UI.  Ignore until we can update the UI to remove this event.
+}
+
+#pragma mark -
+#pragma mark Observers
+
+-(void) prefDisplayIconUpdatedTo:(NSInteger)prefDisplayIcon {
+    // if the pref is set such that the icon is hidden in both the menu bar and the dock,
+    // show the icon in the dock regardless of the user's preference.
+    if (0 == (prefDisplayIcon & (kKCPrefDisplayIconInMenuBar | kKCPrefDisplayIconInDock))) {
+        prefDisplayIcon = prefDisplayIcon | kKCPrefDisplayIconInDock;
+    }
+    
+    ProcessSerialNumber psn = { 0, kCurrentProcess };
+    if (prefDisplayIcon & kKCPrefDisplayIconInDock) {
+        // show dock icon
+        TransformProcessType(&psn, kProcessTransformToForegroundApplication);
+    }
+    else {
+        // hide dock icon
+        preferencesWindow.canHide = NO;
+        TransformProcessType(&psn, kProcessTransformToUIElementApplication);
+    }
+
+    if (prefDisplayIcon & kKCPrefDisplayIconInMenuBar) {
+        // show icon in menu bar
+        [self createStatusItem];
+    }
+    else {
+        // hide icon in menu bar
+        [self deleteStatusItem];
+    }
+}
+
+-(void) observeValueForKeyPath:(NSString *)keyPath ofObject:(id)object change:(NSDictionary<NSKeyValueChangeKey,id> *)change context:(void *)context {
```

Comment:
It appears that NSKeyValueChangeKey is only available in the latest Xcode / macOS SDK 10.12+

If we want to maintain compatibility with "latest tools version minus one" I suggest we omit this from the method signature.